### ISSUE TITLE
:dump does not work properly with cyrillics

### ISSUE THREAD

**Issue by bodqhrohro** [luakit/luakit#768]
## I'm submitting a…
Bug Report

## Current Behavior
Go to https://www.linux.org.ru/. Enter `:dump`. The suggested filename is `LINUX.ORG.RU___________________________________________________Linux_-_https___www.linux.org.ru_.html`, and the encoding of the saved page is broken.
```
LINUX.ORG.RU
Ð ÐµÐ³Ð¸ÑÑ‚Ñ€Ð°Ñ†Ð¸Ñ - Ð’Ñ…Ð¾Ð´
ÐÐ¾Ð²Ð¾ÑÑ‚Ð¸
Ð“Ð°Ð»ÐµÑ€ÐµÑ
Ð¤Ð¾Ñ€ÑƒÐ¼
Ð¢Ñ€ÐµÐºÐµÑ€
ÐŸÐ¾Ð¸ÑÐº
...
```

## Expected Behavior

Properly work with cyrillics, both in filenames and in content.

## Reproduction Instructions

See "Current Behavior".

## Environment

Linux Distribution & Version: Debian GNU/Linux 10.
Output of `luakit --version`: luakit 2.0.0
  built with webkit 2.22.6 (installed version: 2.22.6)

**Comment by bodqhrohro** [comment]
Important note: it turned out that this site does not have an explicit charset meta tag, the charset is only specified in HTTP header. So the issue is mostly on the site engine side. But anyway, it would be useful to assume UTF-8 as the default charset in LuaKit, like other browsers do, or to auto-detect charsets better. BTW, I couldn't find a command for manual charset switching as well.

**Comment by aidanholm** [comment]
You're right that autodetection should be better; that's something that'd be difficult to do from luakit, however.

Manually switching the charset should be possible to add. There's `webkit_web_view_set_custom_charset()` which could be wrapped as a webview widget property, and then set from a lua command. This would be a fairly straight-forward feature to add.

You can in fact set the default charset: it's the `webview.default_charset` setting on [luakit://settings/](luakit://settings/), and is only set to ISO 8859-1 by default because that's the webkit default; presumably for compatibility reasons. I have it set to `UTF-8` with no issues.

**Comment by bodqhrohro** [comment]
>something that'd be difficult to do from luakit, however

Why? There are a lot of libraries for it. Or you mean the difficulity of intercepting the downloaded HTML and changing the charset after the request is completed, but before it is rendered by WebKit?

>it's the webview.default_charset setting on luakit://settings/

Great, thanks!

**Comment by aidanholm** [comment]
> Why? There are a lot of libraries for it. Or you mean the difficulity of intercepting the downloaded HTML and changing the charset after the request is completed, but before it is rendered by WebKit?

Webkit doesn't offer any mechanism or API for intercepting content between the fetch and render stages; in particular, requests are almost always not complete before rendering starts, so that approach won't work. Luakit could (if the custom charset API mentioned above were exposed to lua) analyze the page contents after the page has loaded and adjust the setting, but this would force a reload of the page.
A more adventurous approach would be to parse and rewrite all local html files, inserting a charset meta element as appropriate.

While that page does send a charset header, the header is of course not sent for a saved HTML file. This is a reasonably well-known issue, so standard practice is to use both a header and a charset meta element. I did some experiments with that site and webkit doesn't seem to do any autodetection of the charset at all, but even if it did, there's a large blob of javascript before the first non-ascii byte. Since IIRC web browsers use only the first few hundred bytes for charset detection (to reduce latency), it's really a website issue in this case.

> Great, thanks!

You're welcome!
